### PR TITLE
Document TokenProvider class

### PR DESCRIPTION
**Pull request by sonalivdeshpande**
This is a draft PR for:

1. Documenting on how to build your very own token provider class
2. Rename FRS -> Azure Fluid Relay service

### CODE REVIEW

File: docs/content/docs/build/auth.md [hunk 1]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
+    private readonly user?: Pick<AzureMember, "userId" | "userName" | "additionalDetails">,
+  );
+
+  fetchOrdererToken(tenantId: string, documentId: string, refresh?: boolean): Promise<ITokenResponse>;
+
+  fetchStorageToken(tenantId: string, documentId: string, refresh?: boolean): Promise<ITokenResponse>;
+}
+```
+
+`fetchOrdererToken` and `fetchStorageToken` are responsible for fetching the orderer and storgae url from the host respectively. They return `TokenResponse` object representing token value along with flag indicating whether token came from cache.
```
Comment (andre4i) [inline docs/content/docs/build/auth.md:123]:
```suggestion
`fetchOrdererToken` and `fetchStorageToken` are responsible for fetching the orderer and storage url from the host respectively. They return `TokenResponse` object representing token value along with flag indicating whether token came from cache.
```

File: docs/content/docs/build/auth.md [hunk 2]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
+          userName: this.user?.userName,
+          additionalDetails: this.user?.additionalDetails,
+      },
+    }).then((response) => {
+        return response.data as string;
+    }).catch((err) => {
+        return err as string;
+  });
+}
+```
```
Comment (sumedhb1995) [inline docs/content/docs/build/auth.md:144]: Let's also consider including some documentation here for what is happening on the server-side with `generateToken`, i.e. within the Azure function. This is so that if the developers choose to use some implementation other than an Azure Function, they will have some directions on how to set that up.

File: docs/content/docs/build/auth.md [hunk 3]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
 
 ## Adding custom data to tokens
 
-{{< placeholder >}}
+You can add custom data such as email id, gender, address, etc for your token generation. See [Connect to an Azure Fluid Relay service]({{< relref "azure-frs.md" >}}) for more information.
```
Comment (sumedhb1995) [inline docs/content/docs/build/auth.md:147]: I think since we are also discussing `additionalDetails` in the code snippet above, we can include a line that says that these values can be added in there

**Review by sumedhb1995 — COMMENTED** [review]
This is looking good! Left some comments but overall general feedback would be to generalize the instructions such that we only refer to the Azure Function as one option and instead focus on explaining the concepts such that the developer can set their code up for any BE architecture they choose to go for

**Review by skylerjokiel — CHANGES_REQUESTED** [review]
I really like the format of these docs but feel they are out of context here. Some thoughts:

1. These should be in their own separate file with a title like: "Tutorial: Writing a TokenProvider with an Azure Function"  
2. These docs should live in DCOM. I believe this doc is already on Shawn's plate to move.
3. I think we still need a TokenProvider doc that is one level higher. Something that abstracts away any specific implementation and simply talks about the principals. I think the higher level doc paired with this example would provide a great overview of this space.

@sdeshpande3 @tylerbutler @sumedhb1995

File: docs/content/docs/build/auth.md [hunk 4]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
 
 ### A secure serverless token provider
 
-One option for building a secure token provider is to create a serverless Azure Function and expose it as a token
-provider. This enables you to store the *tenant secret key* on a secure server. Your application calls the Azure Function to
-generate tokens rather than signing them locally like the `InsecureTokenProvider` does.
+One option for building a secure token provider is to create a serverless Azure Function and expose it. This enables you to store the *tenant secret key* on a secure server. Your application calls the Azure Function to generate tokens rather than signing them locally like the `InsecureTokenProvider` does.
+
+You can create the token provider class say `AzureFunctionTokenProvider` which extends the `ITokenProvider` interface without exposing the tenant key secret in client-side code. This class would be responsible for fetching the token from your very own backend. It accepts the url to Azure function and an optional user object.
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply docs/content/docs/build/auth.md:106]: I would rephrase the "say `AzureFunctionTokenProvider`" part like this:

You can create a token provider class which extends the `ITokenProvider` interface without exposing the tenant key secret in client-side code. In the example below the token provider class is called `AzureFunctionTokenProvider`.

File: docs/content/docs/build/auth.md [hunk 5]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
+
+```typescript
+import { ITokenProvider, ITokenResponse } from "@fluidframework/azure-client";
+
+export class AzureFunctionTokenProvider implements ITokenProvider {
+  constructor(
+    private readonly azFunctionUrl: string,
+    private readonly user?: Pick<AzureMember, "userId" | "userName" | "additionalDetails">,
```
Comment (tylerbutler) [inline docs/content/docs/build/auth.md:114]: Is there a simpler way to express this with a little less "type magic?" No worries if not.

File: docs/content/docs/build/auth.md [hunk 6]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
+  );
+
+  fetchOrdererToken(tenantId: string, documentId: string, refresh?: boolean): Promise<ITokenResponse>;
+
+  fetchStorageToken(tenantId: string, documentId: string, refresh?: boolean): Promise<ITokenResponse>;
```
Comment (tylerbutler) [inline docs/content/docs/build/auth.md:119]: Do these intentionally omit the implementation? If so, I suggest making the code reflect that, and add a comment like `// omitted` so it's clear.

File: docs/content/docs/build/auth.md (hunk 1 above)
Comment (tylerbutler) [inline docs/content/docs/build/auth.md:123]:
```suggestion
The `fetchOrdererToken` and `fetchStorageToken` functions are responsible for fetching the orderer and storage URLs from the host. Both functions return `TokenResponse` objects representing the token value along with a flag indicating whether token came from the cache.
```

Comment (tylerbutler) [reply docs/content/docs/build/auth.md:123]: I think this is the first time you mention the cache. Is it important to note that here? If not, just omit it.

File: docs/content/docs/build/auth.md [hunk 7]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
+}
+```
+
+`fetchOrdererToken` and `fetchStorageToken` are responsible for fetching the orderer and storgae url from the host respectively. They return `TokenResponse` object representing token value along with flag indicating whether token came from cache.
+
+To ensure that the secret doesn't get exposed, it is passed to a secure, backend Azure function from which the token is fetched. One of the possible way to fetch the token is making an axios `GET` request call to your Azure function by passing in the tenantID, documentId and userID/userName as optional parameters. Azure function is responsible for mapping between the tenant ID to a tenant key secret to generate and sign the token such that the service will accept it.
```
Comment (tylerbutler) [inline docs/content/docs/build/auth.md:125]:
```suggestion
To ensure that the tenant secret key is kept secure, it is stored in a secure backend location and is only accessible from within the Azure Function. One way to fetch a signed token is to make a `GET` request to your Azure Function, providing the `tenantID` and `documentId`, and optionally `userID`/`userName`. The Azure Function is responsible for the mapping between the tenant ID and a tenant key secret to appropriately generate and sign the token such that the Azure Fluid Relay service will accept it.
```

File: docs/content/docs/build/auth.md [hunk 8]
```diff
@@ -101,13 +101,50 @@ provider is useful for testing, but in production scenarios you must use a secur
...
+
+```typescript
+private async getToken(tenantId: string, documentId: string): Promise<string> {
+  return axios.get(this.azFunctionUrl, {
```
Comment (tylerbutler) [inline docs/content/docs/build/auth.md:129]: I don't have any problem with axios, but is there a vanilla JS way to do this that isn't ugly or illegible? If not, then I suggest adding a callout above this code sample that says, "The example below uses the axios (add a link to the npm package) library to make HTTP requests. You can use other libraries or approaches to making an HTTP request."

**Review by tylerbutler — CHANGES_REQUESTED** [review]
I really like this! I had some wording suggestions, and I also agree with Skyler that this should be its own top-level article. I suggest moving it to a file called `tokenproviders.md` right next to the auth.md file. Title could be "Secure serverless token providers" (or what Skyler suggested).

We will need to move this to DCOM but we can iterate on it here.

File: docs/content/docs/build/auth.md (hunk 5 above)
Comment (skylerjokiel) [reply docs/content/docs/build/auth.md:114]: For brevity we can just use AzureMember here. :)

```suggestion
    private readonly user?: AzureMember,
```

File: docs/content/docs/build/auth.md (hunk 8 above)
Comment (skylerjokiel) [reply docs/content/docs/build/auth.md:129]: ^^ This is why my suggestion was to have this as a `Recipe` or something talking about how to build a specific scenario. Then we have a more basic doc that talks about the principals where you can sudo code something like:

```js
private async getToken(tenantId, documentId) {
  // make call to backend service to retrieve server generated token 
  const token = getTokenFromServer(...);
  return token;
```

File: docs/content/docs/build/tokenproviders.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+menuPosition: 3
+editor: sdeshpande3
+---
+
+The token provider is responsible for creating and signing tokens that the `@fluidframework/azure-client` uses to make
+requests to the Azure Fluid Relay service. There are two ways by which you can generate the token:
+
+- `InsecureTokenProvider`
+
+- Implement your own `TokenProvider` class for fetching the token and a backend for token generation
```
Comment (tylerbutler) [inline docs/content/docs/build/tokenproviders.md:12]:
```suggestion
- `InsecureTokenProvider`
- Implement your own `TokenProvider` class for fetching the token and a backend for token generation
```

File: docs/content/docs/build/tokenproviders.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+The `ITokenProvider` interface has two token calls: `fetchOrdererToken` and `fetchStorageToken`. They are responsible for fetching the orderer and storage URLs from the host respectively. Both functions return `TokenResponse` objects representing the token value.
+
+There are few possible ways through which you can setup your own backend service for token generation such as a backend API endpoint that will handle the token generation, etc. One of the possible backend service which you can setup is an `Azure Function`. This lets you run your code in a serverless environment without having to create a virtual machine or publish a web application. It allows you to write less code, maintain less infrastructure, and save on costs.
+
+## TokenProvider class example
+In the example below, the token provider class is called `AzureFunctionTokenProvider`. This class would be fetching the token from your very own backend Azure Function. It accepts the URL to your Azure Function and an optional user object.
```
Comment (tylerbutler) [inline docs/content/docs/build/tokenproviders.md:24]:
```suggestion
## TokenProvider class example

In the example below, the token provider class is called `AzureFunctionTokenProvider`. This class would be fetching the token from your very own backend Azure Function. It accepts the URL to your Azure Function and an optional user object.
```

File: docs/content/docs/build/tokenproviders.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+requests to the Azure Fluid Relay service. There are two ways by which you can generate the token:
+
+- `InsecureTokenProvider`
+
+- Implement your own `TokenProvider` class for fetching the token and a backend for token generation
+
+Fluid provides an `InsecureTokenProvider` that accepts your tenant secret, then locally generates and returns a signed token. This type of token provider is only useful for development purpose. For production scenarios, you must use a secure token provider.
+
+## Implementing your own TokenProvider class
+The `TokenProvider` class can be created by extending the `ITokenProvider` interface, thereby not exposing the tenant key secret in the client-side bundle code. While the tenant ID, orderer and storage URLs are fine to be on client-side code, the tenant key itself should not be exposed as it would allow malicious users to gain access to your tenant. The `TokenProvider` class would fetch the token from your very own backend service, thus providing a secure way for token resolution.
```
Comment (tylerbutler) [inline docs/content/docs/build/tokenproviders.md:17]:
```suggestion
## Implementing your own TokenProvider class

The `TokenProvider` class can be created by extending the `ITokenProvider` interface, thereby not exposing the tenant key secret in the client-side bundle code. While the tenant ID, orderer and storage URLs are fine to be on client-side code, the tenant key itself should not be exposed as it would allow malicious users to gain access to your tenant. The `TokenProvider` class would fetch the token from your very own backend service, thus providing a secure way for token resolution.
```

**Review by tylerbutler — APPROVED** [review]
Couple of formatting nits; ship it! :shipit:

**Review by tanviraumi — APPROVED** [review]
:shipit:

File: docs/content/docs/build/tokenproviders.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+The `ITokenProvider` interface has two token calls: `fetchOrdererToken` and `fetchStorageToken`. They are responsible for fetching the orderer and storage URLs from the host respectively. Both functions return `TokenResponse` objects representing the token value.
+
+There are few possible ways through which you can setup your own backend service for token generation such as a backend API endpoint that will handle the token generation, etc. One of the possible backend service which you can setup is an `Azure Function`. This lets you run your code in a serverless environment without having to create a virtual machine or publish a web application. It allows you to write less code, maintain less infrastructure, and save on costs.
+
+## TokenProvider class example
+
+In the example below, the token provider class is called `AzureFunctionTokenProvider`. This class would be fetching the token from your very own backend Azure Function. It accepts the URL to your Azure Function and an optional user object.
```
Comment (sumedhb1995) [inline docs/content/docs/build/tokenproviders.md:25]:
```suggestion
In the example below, the token provider class is called `AzureFunctionTokenProvider`. This class would be fetching the token from your very own backend Azure Function. It accepts the URL to your Azure Function and an optional user object. This specific implementation is also provided for you as an export from the `@fluidframework/azure-client` package.
```

**Review by sumedhb1995 — APPROVED** [review]
:shipit:

File: docs/content/docs/build/tokenproviders.md [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+            status: 400,
+            body: "No documentId provided in query params"
+        };
+    }
+
+    let user = { name: userName, id: userId };
+    if (!userId || !userName) {
+        // generateUser will generate a random userName and userId. If either of these values are not provided as
+        // query parameters, the randomly generated values will be used instead.
+        const generatedUser = generateUser() as any;
```
Comment (sumedhb1995) [inline docs/content/docs/build/tokenproviders.md:124]: I think we should drop the random user generation here and instead expect people to pass in the user object for this production scenario. This will also allow us to avoid having the `@fluidframework/server-service-utils` dependency

**Review by skylerjokiel — APPROVED** [review]
🍁

**Review by tanviraumi — APPROVED** [review]
:shipit:

File: docs/content/docs/build/tokenproviders.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+To ensure that the tenant secret key is kept secure, it is stored in a secure backend location and is only accessible from within the Azure Function. One way to fetch a signed token is to make a `GET` request to your Azure Function, providing the `tenantID` and `documentId`, and `userID`/`userName`. The Azure Function is responsible for the mapping between the tenant ID and a tenant key secret to appropriately generate and sign the token such that the Azure Fluid Relay service will accept it.
+
+```typescript
+private async getToken(tenantId: string, documentId: string): Promise<string> {
+    const params = {
+        tenantId,
+        documentId,
+        userId,
```
Comment (sumedhb1995) [inline docs/content/docs/build/tokenproviders.md:58]:
```suggestion
        this.userId,
```

File: docs/content/docs/build/tokenproviders.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        userName
```
Comment (sumedhb1995) [inline docs/content/docs/build/tokenproviders.md:59]:
```suggestion
        this.userName
```

**Review by sumedhb1995 — APPROVED** [review]
:shipit:
